src/arch/util/machine-smp.C

```diff
@@ -589,21 +531,60 @@ static void CmiDestroyLocks(void)
 
 #if !CMK_SHARED_VARS_UNAVAILABLE
 
+class Barrier {
+public:
+  Barrier(const Barrier&) = delete;
+  Barrier& operator=(const Barrier&) = delete;
+
+  explicit Barrier(unsigned int count) :
+    curCount(count), barrierCount(count), curGeneration(0) {
+  }
+
+  void wait() {
+    std::unique_lock<std::mutex> lock(barrierMutex);
```

Comment:
Lock is released automatically when going out of scope?

Reply:
Yes.